pkg/ring/kv/etcd/etcd.go

```diff
@@ -0,0 +1,187 @@
+package etcd
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/ring/kv/codec"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/flagext"
+	"github.com/etcd-io/etcd/clientv3"
+	"github.com/go-kit/kit/log/level"
+)
+
+// Config for a new etcd.Client.
+type Config struct {
+	Endpoints   []string      `yaml:"endpoints"`
+	DialTimeout time.Duration `yaml:"dial_timeout"`
+	MaxRetries  int           `yaml:"max_retries"`
+}
+
+// Client implements ring.KVClient for etcd.
+type Client struct {
+	cfg   Config
+	codec codec.Codec
+	cli   *clientv3.Client
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet.
+func (cfg *Config) RegisterFlags(f *flag.FlagSet, prefix string) {
+	cfg.Endpoints = []string{}
+	f.Var((*flagext.Strings)(&cfg.Endpoints), prefix+"etcd.endpoints", "The etcd endpoints to connect to.")
+	f.DurationVar(&cfg.DialTimeout, prefix+"etcd.dial-timeout", 10*time.Second, "The dial timeout for the etcd connection.")
+	f.IntVar(&cfg.MaxRetries, prefix+"etcd.max-retries", 10, "The maximum number of retries to do for failed ops.")
+}
+
+// New makes a new Client.
+func New(cfg Config, codec codec.Codec) (*Client, error) {
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   cfg.Endpoints,
+		DialTimeout: cfg.DialTimeout,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		cfg:   cfg,
+		codec: codec,
+		cli:   cli,
+	}, nil
+}
+
+// CAS implements kv.Client.
+func (c *Client) CAS(ctx context.Context, key string, f func(in interface{}) (out interface{}, retry bool, err error)) error {
+	var revision int64
+
+	for i := 0; i < c.cfg.MaxRetries; i++ {
+		resp, err := c.cli.Get(ctx, key)
+		if err != nil {
+			level.Error(util.Logger).Log("msg", "error getting key", "key", key, "err", err)
+			continue
+		}
+
+		var intermediate interface{}
+		if len(resp.Kvs) > 0 {
+			intermediate, err = c.codec.Decode(resp.Kvs[0].Value)
+			if err != nil {
+				level.Error(util.Logger).Log("msg", "error decoding key", "key", key, "err", err)
+				continue
+			}
+			revision = resp.Kvs[0].ModRevision
+		}
+
+		var retry bool
+		intermediate, retry, err = f(intermediate)
+		if err != nil {
+			level.Error(util.Logger).Log("msg", "error CASing", "key", key, "err", err)
+			if !retry {
+				return err
+			}
+			continue
+		}
+
+		if intermediate == nil {
+			panic("Callback must instantiate value!")
+		}
```

Comment:
This needs to match the consul client for HA tracker, which sometimes doesn't want to finish the CAS.

Reply:
Fixed.